core/src/test/scala/unit/kafka/server/AddPartitionsToTxnRequestServerTest.scala

```diff
@@ -55,22 +65,149 @@ class AddPartitionsToTxnRequestServerTest extends BaseRequestTest {
     val producerId = 1000L
     val producerEpoch: Short = 0
 
-    val request = new AddPartitionsToTxnRequest.Builder(
-      transactionalId,
-      producerId,
-      producerEpoch,
-      List(createdTopicPartition, nonExistentTopic).asJava)
-      .build()
+    val request =
+      if (version < 4) {
+        AddPartitionsToTxnRequest.Builder.forClient(
+          transactionalId,
+          producerId,
+          producerEpoch,
+          List(createdTopicPartition, nonExistentTopic).asJava)
+          .build()
+      } else {
+        val topics = new AddPartitionsToTxnTopicCollection()
+        topics.add(new AddPartitionsToTxnTopic()
+          .setName(createdTopicPartition.topic())
+          .setPartitions(Collections.singletonList(createdTopicPartition.partition())))
+        topics.add(new AddPartitionsToTxnTopic()
+          .setName(nonExistentTopic.topic())
+          .setPartitions(Collections.singletonList(nonExistentTopic.partition())))
+
+        val transactions = new AddPartitionsToTxnTransactionCollection()
+        transactions.add(new AddPartitionsToTxnTransaction()
+          .setTransactionalId(transactionalId)
+          .setProducerId(producerId)
+          .setProducerEpoch(producerEpoch)
+          .setVerifyOnly(false)
+          .setTopics(topics))
+        AddPartitionsToTxnRequest.Builder.forBroker(transactions).build()
+      }
 
     val leaderId = brokers.head.config.brokerId
     val response = connectAndReceive[AddPartitionsToTxnResponse](request, brokerSocketServer(leaderId))
+    
+    val errors = 
+      if (version < 4) 
+        response.errors.get(AddPartitionsToTxnResponse.V3_AND_BELOW_TXN_ID) 
+      else 
+        response.errorsForTransaction(response.getTransactionTopicResults(transactionalId))
```

Comment:
nit: `response.errorsForTransaction(response.getTransactionTopicResults(transactionalId))` looks really weird from an encapsulation perspective. Can't we use `response.errors.get(transactionalId)`?